### PR TITLE
Queue fetching previews so each url is only fetched once

### PR DESCRIPTION
**Pull request by crosma**
This solves issue #1614. Rather than fetching a url once per recipient, put each client and msg into an array to be processed after the attempt to fetch completes.

### CODE REVIEW

**Comment by crosma** [comment]
Looks like I missed a bug. Can ditch this PR.

**Comment by crosma** [comment]
Found a bug. Will have an update soon.

File: src/plugins/irc-events/link.js [hunk 1]
```diff
@@ -38,17 +40,32 @@ module.exports = function(client, chan, msg) {
...
-				return;
-			}
+		if (!fetch_recipients[preview.link]) {
+			fetch_recipients[preview.link] = [{
+				msg: msg,
+				client: client,
+			}];
+
+			fetch(preview.link, function(res) {
+				if (res === null) {
```
Comment (xPaw) [inline src/plugins/irc-events/link.js]: You can do not null check and call `parse` inside.

File: src/plugins/irc-events/link.js [hunk 2]
```diff
@@ -11,6 +11,8 @@ const storage = require("../storage");
 
 process.setMaxListeners(0);
 
+const fetch_recipients = {};
```
Comment (xPaw) [inline src/plugins/irc-events/link.js]: Camel case please.

File: src/plugins/irc-events/link.js [hunk 3]
```diff
@@ -38,17 +40,32 @@ module.exports = function(client, chan, msg) {
 	})).slice(0, 5); // Only preview the first 5 URLs in message to avoid abuse
 
 	msg.previews.forEach((preview) => {
-		fetch(preview.link, function(res) {
-			if (res === null) {
-				return;
-			}
+		if (!fetch_recipients[preview.link]) {
+			fetch_recipients[preview.link] = [{
```
Comment (xPaw) [inline src/plugins/irc-events/link.js]: Create the object outside of the `if` so it's not defined in two places.

**Comment by xPaw** [comment]
@crosma I like the idea over all, thanks for doing this! I have a question on how would this handle with 200 IRC connections, I imagine there might be enough latency for the link request to be executed multiple times if the link gets fetched before all IRC connections get the link parsed. Would you be able to test this out?

Also, could you confirm that the preview correctly get stored if you reload the page?

**Comment by crosma** [comment]
Currently there is no caching of the results for these previews. It is definitely the case that the request is repeated if a client requests the link after it has been processed, including both because of latency or a page refresh. Or if the link is shared multiple times.  Also the thumbnail would be written to disk each time.

**Comment by crosma** [comment]
I created a new issue to discuss the caching issue: #1957

**Comment by xPaw** [comment]
Page refresh definitely shouldn't re-fetch the link, I was asking if the info stays visible between page refreshes (`preview` gets updated on the `msg` object).

**Comment by crosma** [comment]
Hey, I was wrong about that, it doesn't fetch links when refreshed.

**Comment by xPaw** [comment]
@crosma Hey, could you look into rebasing this? Master was changed to fetch urls and respect user language to provide more correct previews on services such as Twitter.

This should be changed to group requests per url+language now. I'm thinking language should be avoided in public mode to increase cache hit.

**Comment by astorije** [comment]
> I'm thinking language should be avoided in public mode to increase cache hit.

Are you sure it's a good idea? I feel like public instances are directed at broader audiences than private instances, so that could be pretty bad 😅
At the very least if we don't want to want to cache url+language pairs, we should force language to be English to avoid surprises, but tbh I don't think we need that extra optimization, as it gets in the way of usability.

**Comment by xPaw** [comment]
> Are you sure it's a good idea? I feel like public instances are directed at broader audiences than private instances, so that could be pretty bad
> At the very least if we don't want to want to cache url+language pairs, we should force language to be English to avoid surprises, but tbh I don't think we need that extra optimization, as it gets in the way of usability.

That's what I'm saying, basically use the same behaviour it was before.

**Comment by astorije** [comment]
Not exactly, because what we had before (well, now on `master`) would cache the first language that previews, so you may end up with results in Finnish even though neither your client or server is fetching from there.

But really what I'm saying is... language+url pair is fine for both modes.

**Comment by astorije** [comment]
But why go through the extra pain though? It's not going to burn a server, and we're looking at, what, 15-20 languages in the worst, worst case scenario (like, I'd be surprised if we have more than 5 users with that many languages/users, and if that's the case, they have enough resources on their server/network to deal with it).

It just doesn't make sense to me to over-optimize like this, especially if that ends up being more code (even an extra `if` is something we'll break someday lol) and reduces usability 🤷‍♂️

**Comment by xPaw** [comment]
Multiply that by multiple languages per user (that special allows for percentages per languages and other crazy stuff) and browsers (can have minor differences). The point is to protect server and website owners (it's trivial to abuse if wanted, but luckily nothing stupid has happened yet).

**Comment by astorije** [comment]
Hey @crosma, do you think you'd be able to get back to this PR so we can merge it? :)

**Comment by astorije** [comment]
@xPaw, do you think you'd like to take this over, or should we close it until @crosma stops by this project again?

**Comment by xPaw** [comment]
@astorije Don't have interest in taking this over yet.

**Comment by astorije** [comment]
Fair enough. @crosma, let us know if you're interested again and we'll be happy to help you get this merged :)

**Pull request closed without merging**